### PR TITLE
Add physical device support

### PR DESCRIPTION
**Pull request by JoeSSS**

### CODE REVIEW

**Comment by JoeSSS** [comment]
@BasThomas would be nice to get a review here, as we want to release this version before publishing at Monday morning.

File: Launcher/Classes/Constants.swift [hunk 1]
```diff
@@ -9,6 +9,7 @@ enum Constants {
         static let installSimulatorOrPluginDevice = "Please install a simulator or plug-in your device.".localized
         static let useLocalBuild = "Skipping download. Use a local app version.".localized
         static let notCompatibleWithDeviceType = "not compatible with chosen device type.".localized
+        static let wrongDeviceSetup = "Please, provide your device IP and bundle identifier. It can be done under 'Configure Device' settings.\n".localized
```
Comment (BasThomas) [inline Launcher/Classes/Constants.swift]: Why the `\n` here?

File: Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift [hunk 2]
```diff
@@ -4,19 +4,23 @@ class DeviceSettingsViewController: NSViewController {
...
 
     @IBOutlet weak var textField: NSTextField!
+    @IBOutlet weak var bundleID: NSTextField!
     
     override func viewDidLoad() {
         super.viewDidLoad()
         if let deviceIP = applicationStateHandler.deviceIP {
             textField.stringValue = deviceIP
         }
+        if let bundleIDdata = applicationStateHandler.bundleID {
```
Comment (BasThomas) [inline Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift]: Why do you bind this to `bundleIDdata` and not just `bundleID`?

File: Launcher/Classes/ViewControllers/TasksViewController.swift [hunk 3]
```diff
@@ -414,6 +414,17 @@ class TasksViewController: NSViewController {
             arguments.append("")
         }
         
+        if let deviceIP = applicationStateHandler.deviceIP,
+            let bundleID = applicationStateHandler.bundleID,
+            physicalDeviceRadioButton.state == .on,
+            !deviceIP.isEmpty {
+            arguments.append("export DEVICE_ENDPOINT=http://\(deviceIP):37265")
```
Comment (BasThomas) [inline Launcher/Classes/ViewControllers/TasksViewController.swift]: Is this port (the `37265`, I think it's a port?) always the same? Maybe add a constant for it?

File: Launcher/Classes/Constants.swift (hunk 1 above)
Comment (JoeSSS) [reply Launcher/Classes/Constants.swift]: heh, good question. Didn't have time to investigate. We just need a new line there, as the method that outputs doesn't do it for free. Do you think it will be better to add the \n in the code?

File: Launcher/Classes/ViewControllers/TasksViewController.swift (hunk 3 above)
Comment (JoeSSS) [reply Launcher/Classes/ViewControllers/TasksViewController.swift]: Yes this will be always the same. It is Calabash default port. I will add it to the constants

File: Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift (hunk 2 above)
Comment (JoeSSS) [reply Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift]: Because bundleID is already an outlet name

File: Launcher/Classes/Constants.swift (hunk 1 above)
Comment (BasThomas) [reply Launcher/Classes/Constants.swift]: Yeah, I think it would be a better fit in the code.

File: Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift (hunk 2 above)
Comment (BasThomas) [reply Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift]: Ah, right. I think it would be more readable when we rename the `bundleID` outlet to `bundleIDTextField` (and also rename the vanilla `textField`; what's that?) and then in turn rename the `bundleIDdata` to `bundleID`.

Comment (JoeSSS) [reply Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift]: Done. Don't like `bundleIDTextField: NSTextField` tho. But this is always better than names before ;)

File: Launcher/Classes/Constants.swift (hunk 1 above)
Comment (JoeSSS) [reply Launcher/Classes/Constants.swift]: Done

File: Launcher/Classes/ApplicationStateHandler.swift [hunk 4]
```diff
@@ -15,11 +15,13 @@ class ApplicationStateHandler {
         case debugState = "debugCheckboxState"
         case cucumberProfile = "cucumberProfile"
         case additionalRunParameters = "additionalRunParameters"
+        case deviceIP = "deviceIP"
+        case bundleID = "bundleID"
     }
     
-    var simulatorRadioButtonState: Int {
+    var simulatorRadioButtonState: Bool? {
```
Comment (BasThomas) [inline Launcher/Classes/ApplicationStateHandler.swift]: Why is this optional? `bool(forKey:)` returns a `Bool`: https://developer.apple.com/documentation/foundation/userdefaults/1416388-bool

Comment (JoeSSS) [reply Launcher/Classes/ApplicationStateHandler.swift]: Thanks for pointing this out. It was actually causing a bug for a new user.

**Comment by JoeSSS** [comment]
@BasThomas is it ok to be merged?

File: Launcher/Classes/ApplicationStateHandler.swift [hunk 5]
```diff
@@ -15,11 +15,13 @@ class ApplicationStateHandler {
         case debugState = "debugCheckboxState"
         case cucumberProfile = "cucumberProfile"
         case additionalRunParameters = "additionalRunParameters"
+        case deviceIP = "deviceIP"
+        case bundleID = "bundleID"
     }
     
-    var simulatorRadioButtonState: Int {
+    var physicalRadioButtonState: Bool {
```
Comment (BasThomas) [inline Launcher/Classes/ApplicationStateHandler.swift]: This doesn't seem right. You removed `simulatorRadioButtonState`, and the renamed `physicalRadioButtonState` references `.simulatorRadioButtonState`.

File: Launcher/Classes/Constants.swift [hunk 6]
```diff
@@ -38,9 +39,19 @@ enum Constants {
...
             static let uniqueElements = main.path(forResource: "get_uniq_elements", ofType: .bash)
             static let simulators = main.path(forResource: "get_sim_list", ofType: .bash)
+            static let physicalDevices = main.path(forResource: "get_physical_device", ofType: .bash)
         }
         enum Ruby {
             static let helpers = main.path(forResource: "helpers", ofType: .ruby)
         }
     }
+    
+    enum DeviceType {
```
Comment (BasThomas) [inline Launcher/Classes/Constants.swift]: Not sure if this should be here. I can imagine that this enum will have some more functionality in the future, and I feel it's not strictly just constants.

File: Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import AppKit
+
+class DeviceSettingsViewController: NSViewController {
+    let applicationStateHandler = ApplicationStateHandler()
```
Comment (BasThomas) [inline Launcher/Classes/ViewControllers/DeviceSettingsViewController.swift]: We're using this pretty frequently. I am almost thinking it would make sense to add a singleton here?!

File: Launcher/Classes/ViewControllers/TasksViewController.swift [hunk 8]
```diff
@@ -278,13 +249,23 @@ class TasksViewController: NSViewController {
         }
     }
     
-    func getSimulatorsCommand() {
-        if let launchPath = Constants.FilePaths.Bash.simulators {
+    func getDevicesCommand(ofType type: Constants.DeviceType) {
+        var path: String?
```
Comment (BasThomas) [inline Launcher/Classes/ViewControllers/TasksViewController.swift]: This can be a `let path: String`

File: Launcher/Classes/ViewControllers/TasksViewController.swift [hunk 9]
```diff
@@ -278,13 +249,23 @@ class TasksViewController: NSViewController {
...
+            path = Constants.FilePaths.Bash.simulators
+        case .physical:
+            path = Constants.FilePaths.Bash.physicalDevices
+        }
+        
+        if let launchPath = path {
             let outputStream = CommandTextOutputStream()
             outputStream.textHandler = { text in
-                let filderedText = text.components(separatedBy: "\n").filter { $0.contains("Simulator") }
+                let filteredText = text.components(separatedBy: "\n").filter { !RegexHandler().matches(for: "\\(([^()])*\\) \\[(.*?)\\]", in: $0).isEmpty }
```
Comment (BasThomas) [inline Launcher/Classes/ViewControllers/TasksViewController.swift]: I would maybe do a `== false` in the `filter` instead of the `!` at the beginning. Especially since you have the `matches` (but are actually checking the `isEmpty`), this is pretty easily misread.

File: Launcher/Info.plist [hunk 10]
```diff
@@ -15,9 +15,9 @@
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>1.0</string>
+	<string>1.1</string>
```
Comment (BasThomas) [inline Launcher/Info.plist]: Please do this in a separate PR next time! 😬
